x-pack/plugins/visualization_lens/public/indexpattern_datasource/indexpattern.tsx

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React from 'react';
+import { render } from 'react-dom';
+import { Datasource, Operation, DataType } from '../';
+
+interface IndexPatternPrivateState {
+  query: object;
+}
+
+class IndexPatternDatasource implements Datasource<IndexPatternPrivateState> {
+  private state: IndexPatternPrivateState;
+
+  constructor(state?: IndexPatternPrivateState) {
+    if (state) {
+      this.state = state;
+    } else {
+      this.state = {
+        query: {},
+      };
+    }
+  }
+
+  toExpression() {
+    return `${JSON.stringify(this.state.query)}`;
+  }
+
+  renderDataPanel({ domElement }: { domElement: Element }) {
+    render(<div>Index Pattern Data Source</div>, domElement);
+  }
+
+  getPublicAPI() {
+    return {
+      getTableSpec: () => [],
+      getOperationForColumnId: () => ({
+        id: '',
+        // User-facing label for the operation
+        label: '',
+        dataType: 'string' as DataType,
+        // A bucketed operation has many values the same
+        isBucketed: false,
+      }),
+
+      // Called by dimension
+      getDimensionPanelComponent: (props: any) => (
+        domElement: Element,
+        operations: Operation[]
+      ) => {},
+
+      removeColumnInTableSpec: (columnId: string) => [],
```

Comment:
These feel like pretty different methods from the others. Should TableSpec itself be a class that has methods like these? (I'm generally a fan of data + functions, but if we're going OO, we may as well stick with it).

Reply:
Where we landed is that the TableSpec is owned by the data source, so it makes sense to have manipulation methods on the datasource public API. The OO-ness of this change is something I'm actually going to change in the next update